python/lsst/daf/butler/core/fileTemplates.py

```diff
@@ -531,9 +531,12 @@ def format(self, ref: DatasetRef) -> str:
             output = output + literal + format(value, format_spec)
 
         # Replace periods with underscores in the non-directory part to
-        # prevent file extension confusion.
+        # prevent file extension confusion. Also replace # in the non-dir
+        # part to avoid confusion with URI fragments
         head, tail = os.path.split(output)
-        output = os.path.join(head, tail.replace(".", "_"))
+        tail = tail.replace(".", "_")
+        tail = tail.replace("#", "HASH")
```

Comment:
We probably dont care about speed here (at the number of times this will be called its negligible), and what is more readable probably depends on who reads it, but FWIW, `tail.replace(".", "_").replace("#", "HASH")` is a bit faster (I personally find it easier to grok, but others might not). This is more a side comment and not an explicit request to change.